src/pool/srv_target.c

```diff
@@ -1094,113 +1090,171 @@ ds_pool_put(struct ds_pool *pool)
 	daos_lru_ref_release(pool_cache, &pool->sp_entry);
 }
 
-void
-pool_fetch_hdls_ult(void *data)
+struct cont_srv_open_arg {
+	uuid_t   pool_uuid;
+	uuid_t   cont_hdl_uuid;
+	uint64_t flags;
+	uint64_t sec_capas;
+};
+
+static inline int
+cont_srv_open_one(void *vin)
+{
+	struct cont_srv_open_arg *arg = vin;
+
+	return ds_cont_srv_open(arg->pool_uuid, arg->cont_hdl_uuid, arg->flags, arg->sec_capas);
+}
+
+int
+ds_pool_srv_open(uuid_t pool_uuid, uuid_t cont_hdl_uuid)
+{
+	struct cont_srv_open_arg arg = {0};
+
+	D_ASSERT(!uuid_is_null(cont_hdl_uuid));
+	uuid_copy(arg.pool_uuid, pool_uuid);
+	uuid_copy(arg.cont_hdl_uuid, cont_hdl_uuid);
+	arg.flags     = 0;
+	arg.sec_capas = ds_sec_get_rebuild_cont_capabilities();
+
+	return ds_pool_task_collective(pool_uuid,
+				       PO_COMP_ST_NEW | PO_COMP_ST_DOWN | PO_COMP_ST_DOWNOUT,
```

Comment:
There seems race during reintegrating target which status maybe still `DOWN` or `DOWNOUT` before pool map being refreshed. If server handle is not opened for the target to be re-integrated in time, will that affect the subsequent reintegration? I am not sure, need @liuxuezhao for further confirmation.

Reply:
When a target is in DOWN or DOWNOUT state, the 'ds_pool_child' could be not ready for this collective call.
I think the design of reintegration is that everything (server handles, EC epoch boundary, etc.) should be populated when target is UP, @liuxuezhao , is it correct?

Reply:
But in your patch, the `sp_srv_cont_hdl` is opened by the local logic after `ds_pool_iv_srv_hdl_fetch()`, that is usually once because `srv_hdl_fetched` check and set. So if some target is not `UP` at that time, then it will lost the chance to open such handle?

Reply:
Indeed, let me fix it. Thanks.